documentation/writing_tests.md

```diff
@@ -16,4 +17,400 @@
+
+Let's say we have a test utility that we call `retry.ts`, with the following content:
+
+```ts
+import { delay as coreDelay } from "@azure/core-http";
+
+/**
+ * A simple abstraction to retry, and exponentially de-escalate retrying, a
+ * given async function until it is fulfilled.
+ * @param {() => Promise<T>} target The async function you want to retry
+ * @param {number} delay The delay between each retry, defaults to 1000
+ * @param {number} timeout Maximum time we'll let this lapse before we quit retrying, defaults to Infinity
+ * @param {number} increaseFactor Increase factor of each retry, defaults to 1
+ * @returns {Promise<any>} Resolved promise
+ */
+export async function retry<T>(
+  target: () => Promise<T>,
+  delay: number = 1000,
+  timeout: number = Infinity,
+  increaseFactor: number = 1
+): Promise<any> {
+  const start = new Date().getTime();
+  let updatedDelay = delay;
+  while (new Date().getTime() - start < timeout) {
+    try {
+      return await target();
+    } catch {
+      await coreDelay(updatedDelay);
+      updatedDelay *= increaseFactor;
+    }
+  }
+  return null;
+}
+```
+
+This code is clearly not specifically related to any of our projects. Moving it out into a common project will help it's discoverability, not only for other developers to find it, but also to avoid having to upload this code at all, if there happens to be an already existing tool that can be used for the same purpose.
+
+## Writing test cases
+
+While considering possible differences in the [Shared and reusable code](#shared-and-reusable-code) that tests might have, or where they might be placed in the [Test folder structure](#test-folder-structure), test should be similar in each one of our projects. We can ensure some level of homogeneity through the use of our [Recommended tools](#recommended-tools), though we still need to explore what are the indications on writing each test case.
+
+In this section we will be examining our recommendations of:
+
+- [What a test is actually testing](#what-a-test-is-actually-testing), where we will examine how many different things each test should be mainly doing.
+- [Test titles](#test-titles), where we will examine how to properly phrase what each test is doing, and how to make them easier to discover through pattern matching.
+- [Inner parts of a test](#inner-parts-of-a-test), where we will go through how the bodies of each test case should generally be.
+- [Using conditionals](#using-conditionals), in which we will focus on minimizing the possibility of having tests that might run differently depending on external factors.
```

Comment:
I feel here needs to be a section about "when to retry". We should generally have LRO methods to avoid retries, but some retries might be useful. The main issue that I can think of is that we don't have a shared utility for retries. I could move the one  I have on KeyVault into a repo in test-utils, for example, then comment on when to use it.